concent_api/core/views.py

```diff
@@ -1,223 +1,321 @@
+
+        message_task_to_compute = load(
+            force_report_computed_task.message_task_to_compute,
+            settings.CONCENT_PRIVATE_KEY,
+            client_public_key,
+        )
+        assert message_task_to_compute.task_id == message_cannot_compute_task.task_id
+
+        if current_time <= message_task_to_compute.deadline + settings.CONCENT_MESSAGING_TIME:
+            ack_message             = Message.objects.filter(task_id = message_cannot_compute_task.task_id, type = "MessageAckReportComputedTask")
+            previous_reject_message = Message.objects.filter(task_id = message_cannot_compute_task.task_id, type = "MessageRejectReportComputedTask")
+
+            if ack_message.exists() or previous_reject_message.exists():
                 raise Http400("Received RejectReportComputedTask but AckReportComputedTask or another RejectReportComputedTask for this task has already been submitted.")
-            store_message(data['type'], data['message_cannot_commpute_task'], request.body)
+            store_message(message.__class__.__name__, message_task_to_compute, request.body)
             return HttpResponse("", status = 202)
+        else:
+            raise Http400("Time to acknowledge this task is already over.")
+    else:
+        if hasattr(message, 'TYPE'):
+            raise Http400("This message type ({}) is either not supported or cannot be submitted to Concent.".format(message.TYPE))
+        else:
+            raise Http400("Unknown message type or not a Golem message.")
 
 
 @api_view
 @require_POST
-def receive(_request, _message):
+def receive(request, _message):
     undelivered_message_statuses    = MessageStatus.objects.filter(delivered = False)
-    last_undelivered_message_status = undelivered_message_statuses.order_by('id').last()
-
+    last_undelivered_message_status = undelivered_message_statuses.order_by('timestamp').last()
     if last_undelivered_message_status is None:
         return None
 
-    # Mark message as delivered
-    last_undelivered_message_status.delivered = True
-    last_undelivered_message_status.save()
+    current_time = int(datetime.datetime.now().timestamp())
 
+    client_public_key = decode_client_public_key(request)
     raw_message_data     = last_undelivered_message_status.message.data.tobytes()
-    decoded_message_data = json.loads(raw_message_data.decode('utf-8'))
-    current_time         = int(datetime.datetime.now().timestamp())
+    decoded_message_data = load(
+        raw_message_data,
+        settings.CONCENT_PRIVATE_KEY,
+        client_public_key,
+    )
 
-    if decoded_message_data['type'] == "MessageAckReportComputedTask":
-        if current_time <= decoded_message_data['message_task_to_compute']['deadline'] + 2 * settings.CONCENT_MESSAGING_TIME:
-            return decoded_message_data
+    # Mark message as delivered
+    if last_undelivered_message_status.message.type == "MessageForceReportComputedTask":
```

Comment:
Sometimes you're checking `message.type` and sometimes `isinstance(decoded_message_data, ...)`. Please choose one and use it consistently. I think you should stick to `isinstance()` here.